jaeger-thrift/src/main/java/io/jaegertracing/thrift/internal/senders/HttpSender.java

```diff
@@ -119,13 +134,98 @@ public Builder withAuth(String authToken) {
       return this;
     }
 
+    public Builder withCertificatePinning(String[] sha256certs /* comma separated */) {
+      pins.addAll(Arrays.asList(sha256certs));
+      return this;
+    }
+
+    /**
+     * Enable accepting self-signed certificates. This will only take effect if pinning is provided.
+     */
+    public Builder acceptSelfSigned() {
+      this.selfSigned = true;
+      return this;
+    }
+
     public HttpSender build() {
       if (authInterceptor != null) {
         clientBuilder.addInterceptor(authInterceptor);
       }
+      try {
+        // Just obtain hostname for SSL feature. Failure is OK if plain http is used.
+        final URI uri = new URI(endpoint);
+        String hostname = hostname = uri.getHost();
+
+        if (!selfSigned && !pins.isEmpty()) {
+          // Pinning Certificate issued by public CA
+          for (String cert: pins) {
+            certificatePinnerBuilder.add(hostname, String.format("%s", cert));
+          }
+          clientBuilder.certificatePinner(certificatePinnerBuilder.build());
+        } else if (selfSigned && !pins.isEmpty()) {
+          /* Issued by private CA, OkHttp's pinner is unable to verify the pins.
+           * Instead, check the sha256 hash value by custom verifier. */
+          acceptSelfSigned(clientBuilder, hostname, pins);
+        }
+      } catch (Exception e) {
+        // Won't validate the endpoint name.
+        // True moment comes at Sender's Build.
+      }
       return new HttpSender(this);
     }
 
+    private void acceptSelfSigned(OkHttpClient.Builder clientBuilder, final String hostname, final List<String> pinlist) {
+      try {
+        final TrustManager[] selfSignedServerTrustManager = new TrustManager[] {
+          new X509TrustManager() {
+            private final String subjectCN = "CN=" + hostname;
+            private final List<String> pins = pinlist;
+
+            private boolean check(X509Certificate[] chain) {
+              // Intersection of pins and every cert in this chain
+              for (X509Certificate cert: chain) {
+                String hash = CertificatePinner.pin(cert);
+                for (String pin: pins) {
+                  if (hash.equals(pin)) {
+                    return true;
```

Comment:
Do I read this right, that the pin works for any cert in the chain, instead of "all" certs?

Reply:
Yes, pinning is to "ensures that there is an intersection between the public keys in the computed trust chain and the set of fingerprints associated with that domain".
https://wiki.mozilla.org/SecurityEngineering/Public_Key_Pinning

Reply:
TIL, thanks!